pkg/visitors/insert.go

```diff
@@ -173,17 +289,82 @@ func genInsertCLI(iType InsertType) func(args []string) (uefi.Visitor, error) {
+			if err != nil {
+				return nil, fmt.Errorf("unable to parse file '%s': %w", args[1], err)
+			}
+		case InsertWhatTypePadFile:
+			padSize, err := strconv.ParseUint(args[1], 0, 64)
+			if err != nil {
+				return nil, fmt.Errorf("unable to parse pad file size '%s': %w", args[1], err)
+			}
+			file, err = uefi.CreatePadFile(padSize)
+			if err != nil {
+				return nil, fmt.Errorf("unable to create a pad file of size %d: %w", padSize, err)
+			}
+		default:
+			return nil, fmt.Errorf("what-type '%s' is not supported, yet", whatType)
+		}
+
+		wherePreposition := ParseInsertWherePreposition(args[2])
+		if wherePreposition == InsertWherePrepositionUndefined {
+			return nil, fmt.Errorf("unknown where-preposition: '%s'", args[2])
+		}
+
+		pred, err := FindFileFVPredicate(args[3])
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse the predicate parameters '%s': %w", args[0], err)
+		}
+
+		// TODO: use InsertWherePreposition to define the location, instead of InsertType
+		var insertType InsertType
+		switch wherePreposition {
+		case InsertWherePrepositionFront:
+			insertType = InsertTypeFront
+		case InsertWherePrepositionEnd:
+			insertType = InsertTypeEnd
+		case InsertWherePrepositionAfter:
+			insertType = InsertTypeAfter
+		case InsertWherePrepositionBefore:
+			insertType = InsertTypeBefore
+		default:
+			return nil, fmt.Errorf("where-preposition '%s' is not supported, yet", wherePreposition)
+		}
+
+		// Insert File.
+		return &Insert{
+			Predicate: pred,
+			NewFile:   file,
+			// TODO: use InsertWherePreposition to define the location, instead of InsertType
+			InsertType: insertType,
+		}, nil
+	}
+}
+
 func init() {
-	RegisterCLI(insertTypeNames[InsertFront],
-		"insert a file at the beginning of a firmware volume", 2, genInsertCLI(InsertFront))
-	RegisterCLI(insertTypeNames[InsertEnd],
-		"insert a file at the end of a firmware volume", 2, genInsertCLI(InsertEnd))
-	RegisterCLI(insertTypeNames[InsertDXE],
-		"insert a file at the end of the DXE firmware volume", 1, genInsertCLI(InsertDXE))
-	RegisterCLI(insertTypeNames[InsertAfter],
-		"insert a file after another file", 2, genInsertCLI(InsertAfter))
-	RegisterCLI(insertTypeNames[InsertBefore],
-		"insert a file before another file", 2, genInsertCLI(InsertBefore))
-	RegisterCLI(insertTypeNames[ReplaceFFS],
-		"replace a file with another file", 2, genInsertCLI(ReplaceFFS))
+	RegisterCLI(insertTypeNames[InsertTypeInsert],
+		"insert a file", 4, genInsertFileCLI())
+	RegisterCLI(insertTypeNames[InsertTypeReplaceFFS],
+		"replace a file with another file", 2, genInsertRegularFileCLI(InsertTypeReplaceFFS))
+	RegisterCLI(insertTypeNames[InsertTypeFront],
+		"(deprecated) insert a file at the beginning of a firmware volume", 2, genInsertRegularFileCLI(InsertTypeFront))
+	RegisterCLI(insertTypeNames[InsertTypeEnd],
+		"(deprecated) insert a file at the end of a firmware volume", 2, genInsertRegularFileCLI(InsertTypeEnd))
+	RegisterCLI(insertTypeNames[InsertTypeDXE],
```

Comment:
It's a bit of a shame that InsertDxe is being deprecated, since it was a fast and easy way to add things to the DXE firmware volume without having to specify a search for the DxeCore. I get that it doesn't fit with the 4 argument requirement the others have though. This is fine for now, though we should probably rethink our parser to allow conditional arguments